### ISSUE TITLE
Problems with template.

### ISSUE THREAD

**Issue by vulcan25** [ishackm/Code-Bank#1]
Hello!

I think there is another template involved.  `base.html` which this template overides.  Is there any chance you could add that to this repo as a separate file so that I have the whole picture?

**Comment by ishackm** [comment]
Hi!
The code you just saw now is the base.html, I will upload the index.html

**Comment by vulcan25** [comment]
Perfect thanks.  If you're using Flask I could demonstrate some code I built previousy which allows you to define the link bar items in Flask, and pass them to the template so that it builds the navigation bar dynamically.  Please do upload the `index.html` file though, so I can see this too.

**Comment by ishackm** [comment]
Great. Thanks a lot. I have uploaded the index.html file also. I am using Flask-Bootstrap module on Flask. I have also uploaded the app.py file if you need it.

**Comment by vulcan25** [comment]
Perfect.  Give me a little while, and I'll upload something to demonstrate how to do this properly.

**Comment by ishackm** [comment]
Great. Thanks very much.

**Comment by vulcan25** [comment]
Okay, so here's a really basic example which I've prepped in the time I had available.

This actually uses bootstrap v3.1.1 because my template code from previously was already based on this.

Here's a working example of the flat HTML: http://jsfiddle.net/gzmw1v9q/

You can see that it is responsive, in that it shrinks on mobile, and the dropdowns work.

I noticed you've tried to use the agency theme from startbootstrap.  My example is actually based on a stripped down version of the grayscale theme from start bootstrap.  Obviously an older version hence the 3.1.1 version of bootstrap.

Here's the repo: https://github.com/vulcan25/flask_nav_example

A few things to point out.  When I looked at your templates there seemed to be `<opening>` and `</closing>` tags all over the place.  Particularly `<html>`, `<body>`, etc.  A correct HTML document in its most basic form looks like this:

```
<html>
<head>
    <title>My Website</title>
    <link rel='stylesheet' href='http://example.com/style.css' />
</head>
<body>
    <p>I am the homepage or something.</p>

    <script type='text/javascript' src='http://example.com/script.js'></script>
</body>
</html>
```

You need to make sure the tags are nested correctly.

Also CSS includes (in the `<link>` tag) belong in the `<head>` section.  Obviously before the head section ends with the `</head>` tag.

Some HTML tags don't need a corresponding closing tag, the close statement can be put inside them, such as the break tag to take a new line: `<br />`.  This is also valid syntax for the `link` tag which is why you see the `/>` at the end in the example above.

Javascript included by the `<script>` tags belong at the bottom of the page, ususally right before the `body` is closed with the `</body>` tag.  You shouldn't use the quick-close shorthand for script tags though, they actually need the closing `</script>` even if they have a `src` attribute and don't themselves contain javascript.  Of course this is also valid syntax:

```
<script type='text/javascript'>
// I am some javascript
alert('hello');
</script>
```

Ususally when you have a number of different script tags, you should have the one which includes `jquery` first, because the following scripts usually depend on it.

---

Jump into my [app.py](https://github.com/vulcan25/flask_nav_example/blob/master/app.py) and you'll see how I've built the nav bar.  I didn't actually use the Flask-Bootstrap extension, I just define my own python datastruture, and assign this as a flask config variable (`app.config['NAVBAR']`).

I then define a `context_processor` where I pass the navbar to the template:

```
@app.context_processor 
def inject_dict_for_all_templates():
    """ Build the navbar and pass this to Jinja for every route
    """
    # Build the Navigation Bar
    nav =[]
    for item in app.config['NAVBAR']:
        nav.append(item)
    return dict(navbar = nav)
```

This effects every route, so saves me having to pile the extra variables on individual `render_template` functions for each different route.

Then within the [base template](https://github.com/vulcan25/flask_nav_example/blob/master/templates/base.html#L116-L131) I have some nested for loops which make this data struture into the actual nav bar.

---

I'm going to stop explaining here, because I"m not really sure what your aim is to acheive. 

I'd  err on the side of caution with flask.  depending on the functionality you need there may be easier ways to build this.  If the majority of your site is going to be static, you may wish to go with something like `jekyll` which would give you this same nav-building functionality and supports the same template language as flask (I think - Jinja2).  This would save running a flask server which is overkill for a static site.  If you want some stuff with file uploads then you could still implement flask for this, or you could have the whole thing served by flask anyway- I dunno there's so many different ways to set things up.

**Comment by vulcan25** [comment]
You'll see in my template aswell the `index.html` takes the most basic form:
```
{% extends "base.html" %}

{% block body %}

<p>Home page</p>

{% endblock %}
```

As I'm extending the base template, anything after the `{% block body %}` and before the next `{% endblock %}` replaces the corresponding block tags in the base template. (overides it).

So to create an upload page, you could just start with `index.html` (copy it to `upload.html`) but instead of `<p>Home Page</p>` have something else, which would then be surrounded by the rest of the base template.


---

Where it gets clever though: If you have a look at [this section in the base](https://github.com/vulcan25/flask_nav_example/blob/master/templates/base.html#L11).

This full `{% block head %} {% endblock %}` section sits within the actual html `<head>` tags.  Don't get confused and think that overiding this replaces the `<head>` section.  In actual fact it adds to it.  Essentially this allows me to do the following in a child template:

```
{% extends "base.html" %}

{% block head %}
<link rel='stylesheet' href='http://example.com/style_sheet/for_index.css' />
{% endblock %}

{% block body %}

<p>Home page</p>

{% endblock %}
```

`http://example.com/style_sheet/for_index.css` will only be rendered within the specifc child template now, so that's useful for applying styles which should only affect one page/template.  Of course because it's adding to the `head` section, any other stylesheets included in the base template will still apply to that child.

I've done the same thing with `<title>{% block title %}{% endblock %}</title>`

You could overide this block in any child template and the result would be a custom per-page title.

All you'd need to put is:

```
{% block title %} An actual title {% endblock %}
```

Obviously the correct nesting of tags is important though.  In the *child* template, each block should end before another begins.

**Comment by ishackm** [comment]
Hi, Thanks very much for your feedback but how can I overcome that bug in the navigation bar. I am trying to create a database website so most of the pages would be not static. Is there anything that is wrong with the code below and can you show me how can I fix it please?

```
`<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.1.3/js/bootstrap.bundle.min.js" integrity="sha256-E/V4cWE4qvAeO5MOhjtGtqDzPndRO1LBk8lJ/PR7CA4=" crossorigin="anonymous"></script>
<script src="/static/vendor/jquery/jquery.min.js"></script>
<!-- Plugin JavaScript -->
<script src="/static/vendor/jquery-easing/jquery.easing.min.js"></script>

<!-- Custom scripts for this template -->
<script src="/static/js/agency.min.js"></script>` 
```

**Comment by vulcan25** [comment]
At a glance, as I mentioned previously.  Jquery should be included before the other JS files, as they will be reliant on jquery, so it needs to be loaded first.  Something like:

```
<script src="/static/vendor/jquery/jquery.min.js"></script>

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.1.3/js/bootstrap.bundle.min.js" integrity="sha256-E/V4cWE4qvAeO5MOhjtGtqDzPndRO1LBk8lJ/PR7CA4=" crossorigin="anonymous"></script>

<script src="/static/vendor/jquery-easing/jquery.easing.min.js"></script>

<script src="/static/js/agency.min.js"></script>` 
```

I'd also open up your browser's dev tools once you've made these changes and confirm that none of the static includes you're hosting locally from your `/static` directory are giving a 404.

Another thing.  Probably worth making sure that any bootstrap CSS includes (again which should be in the `head` section, not at the bottom of the page like the JS includes) match the version of your bootstrap JS.

So in the code section right here you've picked 4.1.3.  I'd Just pull the matching CSS version from the CDN, and replace any link to `bootstrap.css` which you have in your `head` section:

```
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.1.3/css/bootstrap.min.css" integrity="sha256-eSi1q2PG6J7g7ib17yAaWMcrr5GrtohYChqibrV7PBE=" crossorigin="anonymous" />
```

---

IF you need to find different versions for yourself ...

As you might have noticed you can load https://cdnjs.com/libraries/twitter-bootstrap/4.1.3 in your browser, which gives you a list of includes for this version.  Don't include that link itself in your template.

To find what you need to paste in, hover at the right of the page and choose COPY -> Copy Link Tag with SRI.  This gives you the block I've just pasted above.  Obviously make sure it has the correct filename, `bootstrap.min.css` in the example above.

- The `min` part means it's minified: all of the whitespace is stripped out, so it transfers faster.
- the SRI part, includes the `integrity` attribute, which is bascically a checksum preventing the include from being tampered in transit, which is obviously a good thing ;)

**Comment by ishackm** [comment]
Great Thank you, after many hours of trying, it finally works! If I have any further questions, is that ok with you?

**Comment by vulcan25** [comment]
I'm happy I could help.  Please feel free to ask further questions 👍

**Comment by ishackm** [comment]
Hi there!

I hope you're well.

I just have a few more questions about Flask, please.

In my website, I have to create a function where the user uploads an excel spreadsheet and then imports the data into matplotlib to create some graphs, which then will be shown on the website.

I have already created an upload function in Flask and also created the matplotlib graphs in Jupyter notebooks. I just need help combining them together in my app.py and also show the images on the website. 

Do you mind helping please and do you know any good resources to achieve this aim?

Many Thanks,

ishack

**Comment by vulcan25** [comment]
One question I have... Which operating system are you using on your development box by the way?

Also can the user's upload the data in CSV instead of EXcel, is that possible?  And what kind of charts are you making, line, bar or something else?

**Comment by ishackm** [comment]
I am using Mac OSX

Yes, the user can upload in the csv file format and the graphs I have made is mostly bar charts  and some pie charts. I have also upload app v2.py for you to see.

**Comment by vulcan25** [comment]
Thanks, you might be interseted to take a look at my personal project, which is to log temperature data submitted from a raspberry pi.  I used the ChartJS library to begin with which can be seen here:

https://ser.v25.uk/d0ac059c370ef464

This has support for a wide variety of charts, including Pi, bar etc: https://www.chartjs.org/samples/latest/

However I found that with a large number of datapoints, this started to cause problems (slow rendering time of the chart) so I also looked at Dygraph, which only supports line graphs, but handles large ammounts of data in a better fashion.  The dygraph version can be seen here:

https://ser.v25.uk/d0ac059c370ef464/dy

The documentaion for this one is here: http://dygraphs.com/

However as you can see, I had some problems with the implementation of this (one line falls from the bottom of the chart, which I haven't solved yet).  Dygraph actually loads the information in a CSV format (which my data was not - so I'm still strugling with converting the data from the chartjs compatible format so to speak).

I haven't actually worked with matplotlib or jupyter.  To sumarise the above:

**Chart.js** : varied chart types, and good for small ammounts of data ( < 5000 points per chart)
**Dygraph**: handles large sets of data (millions of points) well, but limited to line charts.

---

As for your implementation there's a few parts to the process, and questions you'd need to ask yourself...

- File uploads, and storing that data somewhere: In a database, or just keep the file on the file system.  The latter is probably easier to begin with, but a database is probably the correct way to lay this out long term.

- Dygraph displays would be really easy with the data already in CSV format.  Perhaps you'd add some validation, so that a user who uploaded a CSV in an incorrect format would be given a warning, rather than it reaching the chart rendering function, and causing errors in the chart redner process.  You really need to be careful accepting file uploads, and some validation needs to be in place.

- Profit?



If you're using mac os, then that's great.  If you haven't already you could download the git client https://git-scm.com/download/mac and make an acutally repository with all your code which you could push to github (unless you have an issue with that?).  This would help me out if I could see all the code you'd written, rather than just one off text file uploads.  IF you need some guidance on how to do that, please let me know.

Any thoughts?

**Comment by vulcan25** [comment]
Of course if you had the matplot lib code developed, and modular enough that it could take a CSV file, and output an image file, all you'd need to do is output the image to the filesystem, and then have flask display said images on the site.  This wouldn't have the interactive (pretty) features of the other chart libs though.  Again if I could see the whole code, then I could give better advice.

**Comment by ishackm** [comment]
Hi, Thanks for the info.

Unfortunately, we are only supposed to use matplotlib for this project and the data points are 10,000 plus. Also, I am making a lot of bar charts so I think matplotlib would be more suitable. Right now I'm testing the matplotlib code on Jupyter and I hope to upload it tomorrow. 

Many Thanks,

Ishack

**Comment by vulcan25** [comment]
Sounds good, look forward to seeing it.

**Comment by ishackm** [comment]
Hi, I have uploaded the templates and the matplotlib script. I am trying to run app v2.py but its giving me an error. I would like to combine it with the matplotlib code so that the when the user uploads a csv it creates graphs and then shows them on the website, please? 

What's the best way to achieve this aim?

**Comment by vulcan25** [comment]
Can you send up the sample CSV file also so I can run this code?

**Comment by ishackm** [comment]
Sure it's done sample.csv

**Comment by vulcan25** [comment]
Also, Did you manage to run this code yourself (the matplot lib part) if so what was the output of this, can you provide a screen shot or similar of what the results of this code were on your machine?  Preferably with the same sample.csv inputted..

Also there are a number of other csv files in the code, can you explain what their functionality is.  Do these come from the user or already exist on the system?

```
9:filename = input("Upload your own data: "); #az20.csv (must be a csv file, flask does not support tsv)
39:uniprot_kinases=pandas.read_csv("https://www.uniprot.org/docs/pkinfam.txt", sep="\t", names=["Column1"],skiprows=74, skip_blank_lines=True )
62:protein_alt_names=pandas.read_csv("Protein_alt_names.csv", header=None)
67:kinase=open("kinase_info_final.csv","r")
102:newfile=open("Protein_alt_names.csv","w+")
119:info4.to_csv("az_filtered.csv", sep=",")
```

**Comment by ishackm** [comment]
`39:uniprot_kinases=pandas.read_csv("https://www.uniprot.org/docs/pkinfam.txt", sep="\t", names=["Column1"],skiprows=74, skip_blank_lines=True )`

This is a csv parsed from online

`62:protein_alt_names=pandas.read_csv("Protein_alt_names.csv", header=None)`

I have uploaded this csv

`67:kinase=open("kinase_info_final.csv","r")`

I have also uploaded this file

All these csv files conduct data analysis of the orginal az20.csv and produce graphs from matplotlib.

The graphs produced:

![img-20190129-wa0007](https://user-images.githubusercontent.com/45857636/52021484-42ea2100-24ed-11e9-90d2-f7d94132f5c7.jpg)
![img-20190129-wa0008](https://user-images.githubusercontent.com/45857636/52021485-42ea2100-24ed-11e9-82e4-703460d2260e.jpg)

**Comment by vulcan25** [comment]
is `az20.csv` == `sample.csv` on your upload?

I only get the one graph outputed

![1](https://i.imgur.com/bxeQjYL.png)

Is this what should be expected when inputing this sample.csv at the prompt?

**Comment by ishackm** [comment]
Yes that's exactly what the output should look like.

is az20.csv == sample.csv on your upload? Yes that's right

**Comment by vulcan25** [comment]
Thanks,

What I'm trying to understand in this code....

The graph I posted to you appears to be generated here, around line 128 which is the last in this block:

```
substrate = info4['substrate']
fold_change = info4['az20_fold_change']
plt.scatter(substrate, fold_change, edgecolors='b')
plt.xlabel('Substrate')
plt.ylabel('Fold Change')
plt.title('Scatter Plot')
plt.show()
```

What about the remaining code that comes after it.?  why is this not executed, and what should it do?

**Comment by ishackm** [comment]
For now, please remove it because for now i just want it to show the image on Flask  after the user uploads the file

**Comment by vulcan25** [comment]
Thanks if it's okay I will take until tomorrow to investigate this fully.  Please stay tuned I think it should work with some small modifications.

**Comment by ishackm** [comment]
Yeah great Thanks very much!

**Comment by vulcan25** [comment]
**DISCLAIMER**: This aint production ready yet.  Consider it a proof of concept.  Probably read this whole thing before attempting.  This has been good to confirm my own learning aswell, it's become a bit of a tutorial.  I've submitted the matplotlib module you provided to my repo (I hope you don't mind, can ice it if you do).  You'll need to import all the txt/csv files yourself to the working directory as I haven't included these.    Again read this whole thing before commencing.

---

A quick word of advice before we start.  This should be possible on a mac if you've got python3.

You can run these commands in a fresh directory:

	python3 -m venv graph
	source graph/bin/activate

Now you have a python virtual environment which keeps anything installed with pip isolated to the graph project. This would be indicated by `(graph)` at the start of your shell. I've packaged a `req.txt` file that contains everything you need to run this.  You can install those requirements with:

	pip install -r req.txt

Then at any point, install more things with `pip install whatever-package` if required.

Then to export what's installed in the venv back to `req.txt` run:

	pip freeze > req.txt

It's the way to go!

As you'll see shortly, the `flask_nav_example` repo from earlier is completely updated, and works minimally.  I strongly suggest you create a new venv from the included `req.txt` and treat this separately from what you've been working on.

I also had to rename `matplotlib.py` to `mat.py` because I installed the module `matplotlib` which would clash if I have a python file of the same name (.py)

---

How I solved this...

I could see that `plt.show` generates the graph, and as you mentioned, commented out everything below this as it's irreverent.

Quick google search for `plt show output png` reveals:

https://stackoverflow.com/a/9888817/2052575

This must be commented out:

	plt.show()

And replaced with:
	
	plt.savefig('foo.png')

`savefig()` is  a matplotib function which can also be passed `foo.pdf` to change the output file format, but we'll stick with `png` format for now.

I run the code again and at the prompt feed it the `sample.csv`  Then confirm that `foo.png`  has been created on the filesystem and contains the expected chart.

This line becomes the end of the code block (everything below is commented)

---

Next I look at the top of the code:

	filename = input("Upload your own data: "); #az20.csv (must be a csv file, flask does not support tsv)

    info=pandas.read_csv(filename, sep="\t")

All we are feeding to this is the filename. So we can make the whole block of code more modular by defining a function, and indenting everything below, down to `plt.savefig()` so it becomes part of that function.  I also pass in an `output` variable which i then pass to `plt.savefig()` so that when we call `proccess_image` I can define the name of the output file name.  It now looks like this:

    filename = input('Upload your own data:')

    def process_image(filename, output):
    	#
    	# [SNIP] Rest of processing code
    	#
    	plt.savefig(output)

Now I can call the function with `process_image(filename, 'output.png')`

I also comment out the line `filename = input ('upload your data')` as we don't want that interactivity happening when we call this as a module. 

I also downloaded `pkinfam.txt` to the working directory, and changed that line in `mat.py` to:

    uniprot_kinases=pandas.read_csv("pkinfam.txt", sep="\t", names=["Column1"],skiprows=74, skip_blank_lines=True )

This avoids the URL, and also means I can comment out `import urlib3` at the top which will make this run better under flask (we don't want to be grabbing remote URLs on every request.)

---

As for the flask side, I had some problems running your app as the templates crashed again.  So I've just used the code I previously had shown you (`flask_nav_example`).  I added flask-restful support and an upload endpoint that I had previously [tested for someone else](https://stackoverflow.com/questions/54393981/try-to-upload-a-file-from-a-uwp-app-to-flask-restful-web-api-failed#comment95600536_54393981).

I move the csv files (and the `pkinfam.txt` which I had previously downloaded to save including `urlib3`) into my working directory, then I can simply test with curl:

    curl -i -X POST  -F 'file=@sample.csv' "http://localhost:5000/upload" -H 'ContentType: multipart/form-data'

Here's what the upload code looks like at this point: [370ecc548e35ee7b3d7dda58824908bfb8e502eb](https://github.com/vulcan25/flask_nav_example/commit/370ecc548e35ee7b3d7dda58824908bfb8e502eb)

As you can see this just puts the file into the following folder:

    UPLOAD_FOLDER = 'static/img' 

I can confirm that with:

	$ ls static/img 
	received.png

Oh shit, that's actually a CSV which has been called `received.png` cause that's what's hard-coded into the script.  Never mind though we can proceed to sort that out:

---

Let's take a quick look at the [panda.read_csv documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html) to work out what that function accepts.  Remember that's what our `process_image` function we created earlier is sending the filename to.

> *filepath_or_buffer* : `str, path object, or file-like object`
> Any valid string path is acceptable. The string could be a URL. Valid URL schemes include http, ftp, s3, and file. For file URLs, a host is expected. A local file could be: file://localhost/path/to/table.csv.
> If you want to pass in a path object, pandas accepts either pathlib.Path or py._path.local.LocalPath.
> By file-like object, we refer to objects with a read() method, such as a file handler (e.g. via builtin open function) or StringIO.


Sounds hopeful.  Let's see if we can make Flask pass the uploaded file straight to our function.	

I import the function we made earlier like so:

	from mat import process_image

Then a quick edit to the `FileUpload` object:

	#photo.save(os.path.join(UPLOAD_FOLDER, filename))
	process_image(photo,'output_from_flask.png')

remember `photo` here is actually the csv file!

Try the above curl command again, and if you're playing along you'll see lots of matplotlib info scroll past on the flask development server and boom: `output_from_flask.png` exists and it contains the chart!

I'll make some changes (mainly rename that pesky `photo` var to `csv` for clarity) and that's us at commit [c3c8b49c551c7a50ac455d38461ae02b471fcf29](https://github.com/vulcan25/flask_nav_example/commit/c3c8b49c551c7a50ac455d38461ae02b471fcf29)  On the next commit I added the `req.txt`.


PROBLEM:  Just got this far and discovered that even though the photo uploads and saves to the local directory, the Flask application immediately crashes. That's not good, here's the error:

    Python[26803:1005902] WARNING: NSWindow drag regions should only be invalidated on the Main Thread! This will throw an exception in the future. Called from (

Quick search reveals:

https://github.com/processing/processing/issues/5676
https://stackoverflow.com/questions/52178375/flask-function-return-invalidated-on-the-main-thread-error

At a glance could be a Mac Mojave bug.  Don't know if we'd get around this buy using a docker container.  (Hint from the future: this will work).

Get docker for mac: https://docs.docker.com/docker-for-mac/install/

If you're gonna do it this way you don't technically need to make the `venv` locally as I mentioned at the start, as all this will live inside docker.

Here's the `Dockerfile`:

	FROM python:3.7.2-stretch

	WORKDIR /usr/src/app

	COPY req.txt ./
	RUN  pip install --no-cache-dir -r req.txt

	CMD ["python","-m","flask","run","--host=0.0.0.0"]


Let's build the docker image and tag it mat-app-image, then confirm it's there:

	docker build -t mat-app-image .

No errors should come up here.  It should grab all the requirements, etc.  Then check it exists"

	$ docker images
	mat-app-image       latest              c1a5fc3be2c4        9 seconds ago       1.18GB

Sweet.  Only slight issue here is it's a pretty large image.  the `python:alpine` image is much smaller, but because of the dependences (matplotlib, etc) I thought stretch was a less troublesome match for the size.  Now run the docker image:

    docker run --name mat-app-runtime -p 9500:5000 -v `pwd`:/usr/src/app mat-app-image:latest

`--name mat-app-runtime` : Here we're giving the running instance a name
`-p 9500:5000` : maps port 9500 on your machine to 5000 within the container
    
    -v `pwd`:/usr/src/app

This line maps a volume from the current working directory to our app directory within the container.  `pwd` in the backticks just grabs the current working directory.

finally `mat-app-image:latest` refers to the image we built earlier.

You should see the flask development server run at this stage.  IF you want it to go in the background, just `CTRL+C` then you must `docker rm mat-app-runtime` before executing the `docker run` command again, this time with the `-d` option to daemonise it.

Then confirm it's running with `docker ps`:

    bd8b4b088263        mat-app-image:latest   "python -m flask run…"   4 seconds ago       Up 2 seconds        0.0.0.0:9500->5000/tcp   mat-app-runtime


Now from another terminal, hit it with curl, first removing the previous `output_from_flask.png` which would have been created:

	rm output_from_flask.png

	curl -i -X POST  -F 'file=@sample.csv' "http://localhost:9500/upload" -H 'ContentType: multipart/form-data'

result:

	HTTP/1.1 100 Continue

	HTTP/1.0 200 OK
	Content-Type: application/json
	Content-Length: 16
	Server: Werkzeug/0.14.1 Python/3.7.2
	Date: Thu, 31 Jan 2019 13:37:39 GMT

	"file uploaded"

And `output_from_flask.png` should now have been created :D


At this stage I commit the actual `mat.py` and the `Dockerfile`.


You can clone the repo in it's current form from https://github.com/vulcan25/flask_nav_example and do the docker steps above to make it work. Remember to copy in the text/csv files to the root direcotry.

---

I'm gonna stop here just now.   Next steps:

- Move the actual CSV/data files into their own subdirectories
- Create an online form, so we don't have to use `curl`.
- Template it based on your templates.
- Implement way to render other chart types.

**Comment by vulcan25** [comment]
I've just been pondering other ways to impement this.  Obviously in it's current implementation there's some security risks because there's no validation  or authentaction on the upload endpoint (as I said **proof of concept!** .  I've been designing some code to do this, which I haven't pushed yet, but it's always a grey area, they say 'don't re-invent the wheel'.

Think about your user base.  Who are they?  Do they all have github accounts for example?

Could the user's push their source .csv files to a github repo for example, then there might be a way to use the github API from your flask app, so when someone accesses your site it will:

- Look at a predefined set of github repos
- If the image doesn't already exist, download the CSVs and run them through the matplotlib function, to generate the images.
- Ideally it would be designed so as to never to do the matplotlib part on the same CSV twice, which would save processing power.

This means you're not worrying about designing the upload functionality yourself, if you follow me.

Of course if this is just for some internal kind of app them maybe the custom upload system would work, I dunno...

**Comment by ishackm** [comment]
Hi, thanks for your detailed research! Very advanced for a beginner like me! I just read a relatable tutorial so  I was wondering if it would be possible to do the same with matplotlib?

https://github.com/viveksb007/camscanner_watermark_remover 

This project is rather about the functionality than security at the moment.

**Comment by vulcan25** [comment]
Yeah, this code is a pretty similar idea.

Where they run they remove_watermark() function in `app.py` is probably where you want the functionality which is in the `mat.py` module of your code.  

Rougly, you could:

    from mat import process_image

At the top of the script.  And then within this function:

```
def process_file(path, filename):
    remove_watermark(path, filename)
```

Call `process_image()` instead of `remove_watermark`.


This is essentally the stuff I've outlined above.  When I moved on to docker this was due to the bug thrown by running the matplotlib stuff within the flask app on OSX, so I'm not sure if you'll stumble on this yourself.

**Comment by vulcan25** [comment]
Hopefully the stuff I've put above isn't too confusing.  I guess the best way to approach it is at your own rate, and you can always refer back later if needed.  I did just want to point out that making something which is internet accessible has a whole other host (no pun intended) of considerations.  Running the app with `python app.py` for the dev server tends to be replaced by something like `gunicorn`+`nginx` in production, due to performance reasons among other things, which could introduce other bugs not spotted in the development stage.  I haven't even covered that part yet lol.

Good luck to you, and if you have any further queries, please don't hesitate to write!

**Comment by ishackm** [comment]
OK thanks, I will give a try and get back to you soon if I get any errors, if you don't mind

**Comment by ishackm** [comment]
Hi, I edited the Camscanner example and uploaded the folder as _matplotlib upload_ , can you please run it and advise on how I could export the graphs to a particular directory and show it on thee Flask website. I also improved the code of the matplotlib so now you should see two graphs.

**Comment by ishackm** [comment]
Also, its az20.txt rather than csv, the code works better with txt.

**Comment by vulcan25** [comment]
I've had a look at the new file `matplotlib_upload/app.py` and I suspect you haven't fully understood what I wrote before, particularly the part about creating the function `process_images()` and laying that code in `mat.py` out like this:
```
def process_image(filename, output):
	#
	# [SNIP] Rest of processing code
	#
	plt.savefig(output)
```

Maybe I didn't explain the reasoning behind this.  This is quite important so let's disect this application:

```
from flask import Flask, render_template
app = Flask(__name__)

@app.route('/')
def index():
    return render_template('index.html')

if __name__ == '__main__':
    app.run(debug=True)
```
What's actually happening here?

First you import the `Flask` class, and the `render_template` function.  The second line here, creates an **instance** of the Flask class (we assign that to `app` in the following code.  It is true to say that `app` is now our **application object**):
```
from flask import Flask, render_template
app = Flask(__name__)
```
---

Then we define a function called `index`.  We use a `decorator` (signified by the `@` which is a python thing) provided by the Flask library, to add a *route* to this function.  The **route decorator** belongs immedaitely above the `def` statement which defines the function.  The `app.` part at the beginning of the decorator means the decorator applies in the **context** of the `app` object we created earlier.
 ```
@app.route('/')
def index():
    return render_template('index.html')
```
The `'/'` passed in the brakets - the first **argument** to the route decorator - specifies the URL or 'endpoint' which will be mapped to the `index` function. In laymans terms when a user hits the root URL (`/`), the `index` function will run and return the rendered template (`index.html` in this case).

---
At this stage, you have a valid flask application.  You could put these 5 lines in any file (for example `run.py` or `ishackm.py`, and it would be true to say you have a `run` or `ishackm` **module**, containing an **application object** called `app`.

You then have the ability to import the **application object** from somewhere else, using a statement like:

    from ishackm import app

It doesn't matter that, in the original file, the `@app.route` part came after you set `app = Flask(__name__)`  because this isn't procedural.  Wherever you do this import the `app`'s context will have the **route decorator** implied.

This may not seem important at the moment, but in the future it will be, because there are numerous advantages to being able to pass the **application object** between modules.  One practical use of this is when it comes to serving in production.  You'll probably end up using a tool like `gunicorn` which can serve WSGI applications.  It just so happens that `app` (our application object) IS a UWSGI application already!!  Gunicorn would therefor be pointed directly at `ishackm:app`.

Sometimes confusion is added because people put all of this in a file and call it `app.py` which is differnt from the `app` contained inside the object.  People (I'm bad for this aswell) then use the two terms interchangably.  This is made worse by the common usage these days of 'app' as an abbreviation for 'application' where years ago was actually a 'computer program' :D. To to summarise:

- `app.py`, `run.py`, `ishackm.py` are python **modules** containing your code.
- `app` which is defined inside said file, is an **application object**, or a **Flask application object**.  Of course you could assign this variable any other title also.

So where am I going with this?  Well, what would happen if you wrote the first 5 lines and ran `python ishackm.py` on that version of the file?  Nothing!  Because all you've done is defined an application object, and a decorated function.  So that's where the last 2 lines come in:

```
if __name__ == '__main__':
    app.run(debug=True)
```

When you execute this file from the command line, with `python ishackm.py` the interpreter automatically sets the variable `__name__` to `"__main__"` so the `if` statement returns true, and the following line executes.  *Alternatively when you `import ishackm` from somewhere else,`__name__` will be set to `"ishackm"` so the following line won't execute.  

So that makes this a good place to execute the development server in, which is what the `run` method of our application object does.  `debug=True` part is obvious.

Essentially `app.run()` is a method which gives you this output at the terminal:

```
 * Serving Flask app "app" (lazy loading)
 * Environment: production
   WARNING: Do not use the development server in a production environment.
   Use a production WSGI server instead.
 * Debug mode: on
 * Running on http://0.0.0.0:5000/ (Press CTRL+C to quit)
 * Restarting with stat
```

You don't want that happening in production because the development server's crap, but ALSO because the production server (`gunicorn`) is designed to load the **application object** directly, rather than executing the `run` method which writes the above data to `stdout`.


---

So back to `matplotlib_upload/app.py` and what's wrong here.
-----------

The main problem is you've just pasted the matplot logic straight into this file, at the same level as all the Flask code.  What this means is if you lauch the application (either via `python app.py` or using `import app from app`) it will execute this matplot code immediately.

To illustrate this, if I add this line `print ("OH DEAR")` on line 1 of my `app.py`, and run the development server, I'll see:

```
$ python app.py
OH DEAR
 * Serving Flask app "app" (lazy loading)
 * Environment: production
...
```

Similary if I try to import my application object, in this case with the python terminal:
```
>>> from app import app
OH DEAR
>>>
# Point to note, I could actually execute app.run() right here.

```

This is a problem for you, because you don't want the matplot logic to be executed when the application object is imported, or the development server runs.

You want it to be in the `proccess_image()` function, which can be passed specific arguments, so that it can be called modularly when required by your computer program ;-)

That way, when it eventually gets tied to an endpoint, via a route decorator, it can be run on demand when someone uploads an image, rather than once when the program starts.


--

I've pushed some form code to my repo, which has upload functionality implemented in javascript with a spinner: https://github.com/vulcan25/flask_nav_example

I've still to implement the directory organization and frontend display of charts, but I'd say this is probably a good imeplmentation so far, based on the points mentioned above.

**Comment by ishackm** [comment]
Oh I see, so the matplotlib code goes after the process.image() function?

**Comment by vulcan25** [comment]
Can you explain this...

```
    kinase=open("kinase_info_final.csv","r")
    line=kinase.readline()
    alt_names={}

    while line != "":
        line=kinase.readline()
        line=line.split(",")
        if len(line)<2:
            break
        elif line[2] not in alt_names:
            alt_names[line[2]]=[]
            genenames=line[3].split()
            for y in genenames:
                if y not in alt_names[line[2]]:
                    alt_names[line[2]].append(y)
            line2=line[-1].split("(")
            del line2[0]
            for y in line2:
                y2=y.replace(")","")
                y3=y2.replace("\n","")
                while True:
                    if y3[-1]==" ":
                        y3=y3[:-1]
                    else:
                        break
                if y3 not in alt_names[line[2]]:
                    alt_names[line[2]].append(y3)
                    
                    
    for x in alt_names:
        if x=='':
            del alt_names[x]
    kinase.close()


    newfile=open("Protein_alt_names.csv","w+")
    for x in alt_names:
        for y in alt_names[x]:
            newfile.write(x+","+y+"\n")
    newfile.close()
```

What's the origin of `kinase_info_final.csv`

Am I correct in understanding that the logic in the above code processes this and writes it to: `Protein_alt_names.csv`

So would it be correct to say that`kinase_info_final.csv` would never have to be read again once this logic is completed, and to make the code more efficient the end application would only ever require `Protein_alt_names.csv`

Or is `kinase_info_final.csv` subject to change?

**Comment by vulcan25** [comment]
Also:
```
    protein_alt_names=pandas.read_csv("Protein_alt_names.csv", header=None)
    protein_alt_names
```

This block appears above the block in my previous comment, but `protein_alt_names` doesn't appear to be used again in the code.  Unless I'm misunderstanding the significance of the second line here.

It looks like this is just reading the file into panda's and printing the info out?

**Comment by ishackm** [comment]
Hi, the kinase file has all the data on all human kinases and the protein file has all the protein names. So yes the kinase file is only used once. 

If you run the matplotlib code on Jupyter or similiar , it wil make more sense.

**Comment by ishackm** [comment]
Yes thats right, it just prints out all the info

**Comment by vulcan25** [comment]
> Oh I see, so the matplotlib code goes after the process.image() function?

The matplotlib code (your custom logic) goes IN the `process.image()` function.

That's what I meant by this part:
```
def process_image(filename, output):
	#
	# [SNIP] Rest of processing code
	#
	plt.savefig(output)
```

`"Rest of proccessing code"`is what was originally in your `matplotlib.py` file.  This is a more modular way of writing it so that you can call it with:

    process_image( INPUT, 'graph.png')

Where:
- `INPUT` is the file which has been uploaded by flask
- the string `graph.png` is then passed through as `output` resulting in `plt.savefig('graph.png')`

**Comment by ishackm** [comment]
Hi, I have uploaded a new version of the _uploadfinal.py_. Can you have a look, please? I am slowly understanding the concept of what you're trying to say but I can't seem to put it together in code. 

Can you please comment the code if it's not too much and show me how to integrate with _app.py_?

**Comment by vulcan25** [comment]
Received. I've got family over this evening but will get back to you once I
have some time to view, probably within 48 hours, at worst 72h ;-)

**Comment by vulcan25** [comment]
Okay, looking at `uploadfinal.py`:

- Everything from line 40 (`#chemical structure instead of removing dot`) to line 225 (`ax.set_title("Top 20 Control mean vs Treated with Inhibitor mean", fontsize=25)`) **inclusive** needs to be indented by **1 Tab** or **4 spaces**.

My editor is [sublime 3](https://www.sublimetext.com/3) which is good and free, so I can do this by highlighting those lines, and pressing `TAB` (indent) and `SHIFT+TAB` (unindent). To move the whole block at once.

Here's a video of that: https://imgur.com/a/k4s6Yvb

Why is this sinificant...

Indentation in python is very important.  Unlike javascript where curly brackets define stuff that's blocked together.

So, in Python:
```
def hello():
    print (1)
print (2)
```
is not the same as
```
def hello():
    print (1)
    print (2)
```
- In the first block, calling `hello()` from somwhere else, would print `1`
- In the second block, calling `hello()` from somewhere else would print, `1` then `2`.

Also, assuming this code was in a module (a file called `sample.py`)  *then*, in another module you put `import sample`:

- In the first block, as soon as that `import` line runs, `2` would be printed.  You could then do `sample.hello()` which would print `1`
- In the second block, as soon as the `import` line runs, nothing will be printed because both print statements are contained within the function, so to print `1` then `2` you'd call `sample.hello()`

- Where's the `sample.hello()` syntaxt coming from... well if you just wanted to call this (from another module or file) as `hello()` you'd need to do this instead: `from sample import hello`.  Even doing this on the first code block, though would still print `2` as soon as the `import` runs.  That's why it's so important to keep code contained within functions.

To illustrate another way I quickly create a file using the terminal:
```
air >  flask_nav_example git:(master) ✗ cat > sample.py
# A quick way to make a new file
print ('I am not indented')
def business():
    print ('Taking care of business')
^C
```
Then demonstrate with the python shell:
```
air >  flask_nav_example git:(master) ✗ python
>>> from sample import business
I am not indented
>>> business()
Taking care of business
>>> sample.business()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'sample' is not defined
>>> 
```
Can you see why the last execution throws an error?  `sample.business()` can't be called directly here.  For that I should have done `import sample`.  Either way though `I am not indented` will always print the second I do the import.

---


So back to your code.  Once you've indented lines 40 -> 225 it begins to look like this:

![fixed slightly](https://i.imgur.com/mxGJ84J.png)

- At this stage, everything from `Line 38` onwards has become part of the `process_file()` function.
- Remove `Line 38`: `process_image()(INPUT, 'graph.png' )` because this isn't valid syntax at all.
- Remove `Line 4`: `from mat import process_image` because the `process_image` function doesn't even exist anymore, and that import line assumes you're trying to import `process_image` from a file called `mat.py`:  *You could set things up this way though, which is what my `flask_nav_example` code is laid out like: but it's essential that within `mat.py` there's a `process_image` image function, which correctly contains the logic (which is indented due to the reasons above).

So assuming we keep the logic in the main `uploadfinal.py` file, and nuke that Line 4 and 38, here's what it now looks like:

![mod](https://i.imgur.com/IL6kgPM.png)

Looking better, but a couple of problems still exist here:

- `Line 48` needs to go because that's running `input()` which is interactive (gets input from the user when the script is executed on the command line) which is no good in a flask application. And in it's current form we're going to pass `filename` in as an arugument to `process_file` which is what this syntax means:

    def process_file(path, filename):

- That function also takes `path`.  We're going to use `path` to define what to call the rendered image of the graph. oth of these variables can be used within the function.  It means we can call it somewhere using the syntax `process_file('image_of_chart.png', 'sample.csv')` (**Obviously the order is important here.**)

- Using the exact variable name `filename` is convenient here, because your logic code already references `filename` which it expects to be a string like `"sample.csv"`  (previously this was assigned on `Line 48` but by removing that It means whatever we pass in as `filename` automatically carries through to the rest of your code, which is ideal.

So where are we supposed to call this form?  The `index()` function just above.  Currently that line looks just like this:

    process_file(os.path.join(app.config['UPLOAD_FOLDER'], filename), filename)

This ought to look more like this:

    process_file( 'output_test.png', file)

I also comment `Line 31` as it's not required, and we've got something liek:

![modified more](https://i.imgur.com/kNAqv6v.png)

- Note that `file` is an actual file object.  This of that as the data which came from the user.
- We're not using the file object's method `.save` (Line 31) anymore, as we don't actually want to save the CSV file.  We only need the output image saved to disk.  Conveniently we can pass the file object in when we call `process_file()` because the function which eventually receives it (`pandas.read_csv` on line 50) actually supports a file name/path as the first argument which causes it to read from said file path, OR a file object in which case it (`pandas.read_csv`) doesn't go near the disk.  The support for either type is implemented by pandas.

Round about line 167, you need to remove `plt.show()` cause that's again an interactive feature which launches a rendering of the chart in your operating sytem's gui.  Obviously not ideal in a flask application, we want to save that image to disk.  If you search my previous post, for `plt.savefig` you'll see the way this is done, so given that you're passing a string `'output_test.png'` to `process_file()` as the `path` argument this needs to look like:

    plt.savefig(path)

Remember to respect the indentation.  That needs to remain within the `process_file()` function, so should remain indented. 
---

I'm going to stop here.  Hopefully the above helps you, I realise it's not fully complete but I'm at a fork where I'm in danger of rewriting some stuff that's already done correctly in my `flask_nav_example` code.  

Perhaps could you give me some feedback on this post.  I'm hoping that it's made a few things clearer, and you may then be able to look at `flask_nav_example` with more knowledge of what's happening and why.

Particularly there may still be bugs based around the `index` function in this code and the file upload stuff, it can get slightly complicated with all the references to `UPLOAD_FOLDER` etc.  Also I don't have your exact templates for this version.


A few final thoughts of this post:


Currently all this is within the `process_file()` function:

    import pandas
    import urllib3
    import numpy as np
    import matplotlib.pyplot as plt
    import seaborn as sns
    from pandas.plotting import _converter

You don't want that there: it will run these library imports every time `process_file` is called.  These belong at the very top of the file so the stuff's imported once, when the flask application runs.  As it stands this would cause performance issues.

**Comment by ishackm** [comment]
Hey Thanks for the commenting through the code, its much clearer now. Do you mind uploading the code that you showed the previous post? I will experiment with it and get back with you asap.

Thanks,

Ishack

**Comment by vulcan25** [comment]
> show me how to integrate with app.py

Everything in this post assumes you're executing `python uploadfinal.py` because in there you have defined the flask application, and that's where all the routes are so `uploadfinal.app` is a valid Flask application object.

Again with the stuff I've gone over, you'll see why in `flask_nav_example` i put the matplot lib code it's own file called `mat.py` and then in my `app.py` I:

    from mat import process_image

then just call the function with process_image() further down the file.

This has the effect of keeping the maths logic separate from the flask logic.

**Comment by vulcan25** [comment]
> Hey Thanks for the commenting through the code, its much clearer now. Do you mind uploading the code that you showed the previous post? I will experiment with it and get back with you asap.

Hopefully no sync errors between this and what I posted.  
I've tried to keep the line spacing as similar as possible.

https://gist.githubusercontent.com/vulcan25/44419bcb8daa9748be4c9abfcef5e1ed/raw/2bd914db89cf71255763ead22c1f6f99f5569c6b/gistfile1.txt

**Comment by vulcan25** [comment]
Just testing something here, will add more description shortly...

**Comment by ishackm** [comment]
Sure. I just would like to say thanks very much for your help so far, I'm getting to understand Flask bit by bit :)

**Comment by ishackm** [comment]
Good news, the files are linking together, I have used your nav bar example to link _uploadfile.py_ and _app.py_ but still the flask website does not link to the _uploader_ route. I will post images soon to explain better.

**Comment by vulcan25** [comment]
No probs, I'm no expert but have been playing with flask alot recently.

You showed two graphs, and I had a third.

Are all three graphs genearted from the same `sample.csv` or are the csv files required to generate each graph different.  This would contribute to the design so is good to know now.

**Comment by ishackm** [comment]
All the csv files are required to generate the graphs as they contribute to the different graphs.

**Comment by ishackm** [comment]
I have updated _app,py_ and _uploadfinal.py_.  Can you download the whole repo and run app.py, please?. You can see all the files work but not the /uploader route

**Comment by vulcan25** [comment]
Okay, I got that to run properly with some edits.

---

This bit doesn't really work properly.  I can see your thinking here but where you set `app=Flask` in an attempt to then set `app.config['UPLOAD_FOLDER']` can't work because the `app` you've just initialized here is completely separate from the `app` which you initialise in `app.py`.  I may come back to this in a separate post, to explain further.  Delete or comment out in the meantime:

```
# def uploaded_file():
#     UPLOAD_FOLDER = os.path.dirname(os.path.abspath(__file__)) + '/csv/'
#     ALLOWED_EXTENSIONS = {'csv'}
#     app = Flask(__name__, static_url_path="/static")
#     DIR_PATH = os.path.dirname(os.path.realpath(__file__))
#     app.config['UPLOAD_FOLDER'] = UPLOAD_FOLDER
```

---

This needs to go in `app.py`
```
ALLOWED_EXTENSIONS = {'csv'}
```

---

This part has to be moved to `app.py`, as it is required by the `index` function, which we're also about to move:
```
def allowed_file(filename):
    return '.' in filename and filename.rsplit('.', 1)[1].lower() in ALLOWED_EXTENSIONS
```

You must also move the whole `index` function from `uploadfinal.py` to `app.py` and change it to look like this:

```
@app.route('/uploader', methods=['GET', 'POST'])
def uploader():
    if request.method == 'POST':
        if 'file' not in request.files:
            print('Please upload a csv file')
            return redirect(request.url)
        file = request.files['file']
        if file.filename == '':
            print('No file selected')
            return redirect(request.url)
        if file and allowed_file(file.filename):
            #filename = secure_filename(file)
            process_file('output_test.png', file )
            return redirect(url_for('index'))
    return render_template('upload.html')

```
I've changed the function name to `uploader` so it doesn't clash with the `index` function that already exists in here.

I've commented the `secure_filename` line, as this is only required when you are using the filename provided by the user (which is what `file.filename` actually is).  Think of that like if I have a file locally called `Some_specific.csv`.  This exact string is available in `file.filename`.  `file` itself is the file object (the data in this CSV).    However a malicous request could be crafted which would set `file.filename` to something like `../../../../secret`, which if you handle imporperly with `os.path` methods, could be used to access stuff it's not supposed to elsewhere on the server.  `filename = secure_filename(filename)` prevents this  but irrelevant in your case, because you're not using the user submitted filename for anything, you pass the `file` object to `process_file()` and specify `output_test.png` yourself which is what the ultimate file will be called.

`process_file` won't import properly, so you need to make sure this import is present at the top of `app.py`:

    from uploadfinal import process_file

get rid of the other line that looks like this `from uploadfinal import uploaded_file` cause that function doesn't exist any more.

Here's a visual of app.py:

![fixed](https://i.imgur.com/fl0evIT.png)


This appears to be successfully creating the file `output_test.png` when I run it. 

If you get an error or two on launch, I've maybe missed some import lines, so just double check against the visual, which works.

I also had to rename the `matplotlib.py` which was kicking about in the repo to something else ([see this answer](https://stackoverflow.com/a/46261207/2052575)) because it wouldn't let me start the application with that present, although this may be specific to my environment.

**Comment by ishackm** [comment]
Hi, that looks good, thanks very much! However, I cant see the app.py file, can you direct it to me please, so that I can run it?

**Comment by vulcan25** [comment]
Also, the `return redirect url_for('index'))` line immediatley after you call `process_image`...

I chose to redirect this to 'index' after a sucessful upload but this could be changed later.

Once you've got some database stuff implemented, you'll do something along the lines of this within the statement: `if file and allowed_file(file.filename):` 

Make a database entry for the now uploaded image, and return the ID of the newly created row in same statement.  Don't overthink this code, but something along the lines of:
```
 # create a new entry based on the model `Graph` which you'll have written elsewhere.
new_entry = Graph(title='Some title')
# Add that to the database session, then flush which automatically sets  `new_entry.id`
# to the id which is assigned by the database
db.add(new_entry)
db.flush()
# new_entry.id is the id of the new entry
```
You would then call:

    process_image(str(new_entry.id)+'_image.png', file)

Which names the actual to something like `8_image.png` which is in turn accossiated with any other data you store regarding that upload, based on `id`.

You'd probably then need to do:

        db.commit()

Which writes anything that went through `db.add()` to the actual database.

Finally, you could `return redirect url_for('view_chart', chart_id = new_entry.id ))`

And write the corresponding route to handle that:

```
@app.route('chart/<int:chart_id>', methods=['GET'])
def view_chart( chart_id ):
    try:
        graph = Graph.query(id = chart_id).one()
        return render_template('graph_display.html', chart=graph)
    except:
        abort(404)
```
Then in the template somethign along the lines of:
```
{% if chart %}
<h1>{{chart.title}}</h1>
<img src='images/{{chart.id}}_image.png' />
{% endif %}
```

This means as soon as a user uploads a chart, they'll be taken to the page displaying that chart.  This isn't perfect but I think illustrates the logic.

**Comment by vulcan25** [comment]
> Hi, that looks good, thanks very much! However, I cant see the app.py file, can you direct it to me please, so that I can run it?

https://i.imgur.com/fl0evIT.png Or did you mean the full file: I've revised the earlier gist:
https://gist.github.com/vulcan25/44419bcb8daa9748be4c9abfcef5e1ed/0c1daf7ba5c987fa61a00f00a82494b7867b7d48

**Comment by ishackm** [comment]
sorry I meant, the whole repository please

**Comment by ishackm** [comment]
Hi, the code works but I get this in the terminal 

`Ishacks-MacBook-Pro:Flask_Bootstrap_final Ishackm$ python app.py 
Upload your own data: 
`

I thought it would open up the index.html and the user clicks on the upload file page and upload their dataset?

**Comment by vulcan25** [comment]
Can you push the changes that you've got up to present.

If you're seeing `upload your own data` when you run `python app.py` most likely the call to the `input()` function has crept back in somehow.

Perhaps check for lines `importing` old stuff.

`python app.py` should run the development server which you can then access via `http://localhost:5000/` in your browser.

**Comment by ishackm** [comment]
oh yes sorry. Now I got a new error

`File "app_terminal.py", line 5, in <module>
    from uploadfinal import process_file
  File "/Users/Ishackm/Desktop/Flask_Bootstrap_final/uploadfinal.py", line 49, in <module>
    info=pandas.read_csv(filename, sep="\t")
NameError: name 'filename' is not defined`

How can i fix this please?

**Comment by vulcan25** [comment]
I'd need to see the current version of `app_terminal.py` can you push all changes?

**Comment by ishackm** [comment]
ok sure

**Comment by ishackm** [comment]
I have uploaded app_terminal.py and the  updated version of uploadfinal.py

**Comment by vulcan25** [comment]
I think you've added a bad version of `uploadfinial.py` as the `app` object is still built in there, and the indentations wrong, also  ` process_image()(INPUT, 'graph.png' )` has returned.  It should look more like this:

https://gist.github.com/vulcan25/4581610cea5099ef124c0e33ae6cc000

**Comment by ishackm** [comment]
Great Thanks! It works finally! 
I just found that I have to produce more graphs but don't worry, the matplotlib code will be easier to understand :)  I will upload it tomorrow.

Is it easy to show the generated graph on the upload2.html page?

Also, the code right now produces one graph but how do I modify the code so that It produces 6 graphs and show them in upload2.html please?

**Comment by ishackm** [comment]
Hi, sorry its 4 graphs

**Comment by ishackm** [comment]
Hi, are you available?

**Comment by ishackm** [comment]
Hi Vulcan, are you available?

**Comment by ishackm** [comment]
I need help linking a database with Flask please?

**Comment by ishackm** [comment]
I have managed to get the file on Flask.

**Comment by vulcan25** [comment]
Can you push the latest version of your code and I'll have a look?

**Comment by ishackm** [comment]
Hi, I am  using this tutorial series:

https://www.blog.pythonlibrary.org/2017/12/12/flask-101-adding-a-database/

I have already created a database

**Comment by ishackm** [comment]
Hi, I have uploaded the new website, `website2`. 

Please run `app_terminal_v2.py` to use the website

> final.db

this is the database file.

When I search in the search box, instead of saying no results, it redirect back to the index.html page

![app](https://user-images.githubusercontent.com/45857636/52365708-fa2bee00-2a3f-11e9-8990-8bbce50c2352.png)

As you can see it does not go to the `@app.route ('/results') ` and stops at the the line before, which redirect back to the index.html

**Comment by vulcan25** [comment]
Nice work, love the template, looks like you've made good progress!

I need to re-familiarize myself with WTF and will get back to you in the next hour or so...

**Comment by vulcan25** [comment]
What's the search box at the top actually meant to do?  What's it meant to search for?

If it's items in the different sections, Kinase, Inhibitor, etc, I noticed you've got datatables activated which includes a search box.  If you are intending on building a search which queries this database, I strongly suggest you just have the visitors utilise the existing datatables search box to filter these results.

You're at a fork:

- Implement your own search in flask, which requires validating user input and bulding queries for all the separate columns in the databases. 

OR:

- Use the datatables built in search which might become an issue if there's tonnes of data, as it will all need to be downloaded on page render, but it looks like this is happening anyway (which isn't a major issue unless there's gonna be gigs of data or something).

---

I think the actual issue that you were describing is that in the HTML your form needs an action, so something like:

Line 62 of `base_index.html` : `  <form method=post action='/results'>`

Then handle that response in the `search_results` function, not `index`.  Something like:

```
@app.route('/', methods=['GET', 'POST'])
def index():
    search = KinaseForm(request.form)

    return render_template('index.html', form=search)

@app.route('/results', methods=['POST'])
def search_results():
    if request.method=="POST":
        #handle results
```

I haven't expanded the logic in this last function too much as you have a better understanding of the data.

**Comment by ishackm** [comment]
Thank you


Hi, I have seemed to fix the issue but I can't seem to load the data from the database; I have to use Flask SQLAlchemy as it is a requirement of the project. AlsoI can't seem to connect to the database so when user searches it shows a blank page

**Comment by vulcan25** [comment]
push code, with changes ;-)

**Comment by vulcan25** [comment]
Noticed you're including `app` from `db.py`.  You want to use the application factory pattern... two secs and I'll update ya...

**Comment by ishackm** [comment]
I have pushed the latest updates- please run app_terminal_v2.py

- Now when you press search on the search bar, it will display the whole `Alt_name` table but when I search for a kinase it goes to the error page.

please see  `model.py`for the whole database tables 

What want to the function to do:

The user searches for a kinase:

It should show the following things:
 The Kinase Name and its following attributes: 
- Gene Symbol 
- Gene_family
- Cell_location

**Comment by vulcan25** [comment]
Check everything that's going on here.

https://github.com/vulcan25/Code-Bank/commit/d70466de8775f95ef7a2bdc2c7e724a16ef6ce8e


I've added a few new files.   `run.py` being the main one, which mean's you would now execute the application with `python run.py`.  This introduces the application factory pattern which you'll thank me for later.

I've changed it so that `app` contained within `app_terminal_v2.py` is now a blueprint which gets registered in the `create_app()` function of `run.py`.  All the routes you've already defined come with it.

database is initialized in `database.py` and the `models.py` file changed slightly to reflect this.

configuration now lives in `config.py`.

You'll probably need to modify the templates with this layout as they will complain about the `url_for` function not leading to the correct view.

Basically the modification in all the following is for each line: `{{ url_for('index') }}` gets replaced with `{{ url_for('main_bp.index') }}` for all of these:

```
$grep -n url_for templates/*.html
templates/base_error.html:36:          <a class="navbar-brand js-scroll-trigger" href="{{ url_for('index') }}">RFJI Protein Database</a>
templates/base_error.html:44:                  <a class="nav-link js-scroll-trigger" href="{{ url_for('kinase_db') }}">Kinase DB</a>
templates/base_error.html:46:                    <a class="nav-link js-scroll-trigger" href="{{ url_for('inhibitor_db') }}">Inhibitor DB</a>
templates/base_error.html:48:                      <a class="nav-link js-scroll-trigger" href="{{ url_for('upload') }}">Upload File</a>
templates/base_index.html:41:                  <a class="nav-link js-scroll-trigger" href="{{ url_for('main_bp.upload') }}">Upload File</a>
templates/base_index.html:43:                    <a class="nav-link js-scroll-trigger" href="{{ url_for('main_bp.kinase_db') }}">Kinase DB</a>
templates/base_index.html:45:                    <a class="nav-link js-scroll-trigger" href="{{ url_for('main_bp.inhibitor_db') }}">Inhibitor DB</a>
templates/base_inhibitor_db.html:52:                <a class="nav-link js-scroll-trigger" href="{{ url_for('index') }}">Home</a>
templates/base_inhibitor_db.html:54:                  <a class="nav-link js-scroll-trigger" href="{{ url_for('upload') }}">Upload File</a>
templates/base_inhibitor_db.html:56:                    <a class="nav-link js-scroll-trigger" href="{{ url_for('kinase_db') }}">Kinase DB</a>
templates/base_kinase_db.html:54:                <a class="nav-link js-scroll-trigger" href="{{ url_for('index') }}">Home</a>
templates/base_kinase_db.html:56:                  <a class="nav-link js-scroll-trigger" href="{{ url_for('upload') }}">Upload</a>
templates/base_kinase_db.html:58:                    <a class="nav-link js-scroll-trigger" href="{{ url_for('inhibitor_db') }}">Inhibitor Database</a>
templates/base_result.html:43:                  <a class="nav-link js-scroll-trigger" href="{{ url_for('index') }}">Home</a>
templates/base_result.html:45:                    <a class="nav-link js-scroll-trigger" href="{{ url_for('kinase_db') }}">Kinase DB</a>
templates/base_result.html:47:                    <a class="nav-link js-scroll-trigger" href="{{ url_for('inhibitor_db') }}">Inhibitor DB</a>
templates/base_sql.html:42:                  <a class="nav-link js-scroll-trigger" href="{{ url_for('upload') }}">Upload File</a>
templates/base_sql.html:44:                    <a class="nav-link js-scroll-trigger" href="{{ url_for('kinase_db') }}">Kinase DB</a>
templates/base_sql.html:46:                    <a class="nav-link js-scroll-trigger" href="{{ url_for('inhibitor_db') }}">Inhibitor DB</a>
templates/base_upload.html:43:                  <a class="nav-link js-scroll-trigger" href="{{ url_for('index') }}">Home</a>
templates/base_upload.html:45:                    <a class="nav-link js-scroll-trigger" href="{{ url_for('kinase_db') }}">Kinase DB</a>
templates/base_upload.html:47:                    <a class="nav-link js-scroll-trigger" href="{{ url_for('inhibitor_db') }}">Inhibitor DB</a>
templates/base_upload_page2.html:43:                  <a class="nav-link js-scroll-trigger" href="{{ url_for('index') }}">Home</a>
templates/base_upload_page2.html:45:                    <a class="nav-link js-scroll-trigger" href="{{ url_for('kinase_db') }}">Kinase DB</a>
templates/base_upload_page2.html:47:                    <a class="nav-link js-scroll-trigger" href="{{ url_for('inhibitor_db') }}">Inhibitor DB</a>
templates/upload.html:31:<a class="btn btn-primary" href="{{ url_for('upload2') }}" role="button">Click Here To Upload </a>
```

I edited the base_index already to make sure IT DOES render.

Any queries please ask.

**Comment by vulcan25** [comment]
with that structure you can initialize the database from the python terminal:
```
$ python3
>>> import database
>>> database.init_db()
```

**Comment by vulcan25** [comment]
This sorta db setup is documented here: http://flask.pocoo.org/docs/1.0/patterns/sqlalchemy/

Any there's a less flask related description here: https://docs.sqlalchemy.org/en/latest/orm/contextual.html#using-thread-local-scope-with-web-applications

The second link has a better explanation of what the scoped session is actually doing in relation to web requests.

**Comment by ishackm** [comment]
> show the following things:
> The Kinase Name and its following attributes:
> 
>     * Gene Symbol
> 
>     * Gene_family
> 
>     * Cell_location

So now I can do this?

**Comment by ishackm** [comment]
Hi, I ran run.py but I can't seem to search nothing in the search box, it all comes back to index.html page. Also I want to initalise the db automatically when i run run.py

**Comment by ishackm** [comment]
![img-20190201-wa0014](https://user-images.githubusercontent.com/45857636/52384128-0d58b100-2a74-11e9-9d26-de664241735a.jpg)

This is how I want it to look when a user searches for a kinase

**Comment by vulcan25** [comment]
I'm out of time for just now, but added another commit, I'm kinda closer, but still bugs out.  

https://github.com/vulcan25/Code-Bank/commit/ede0389e5d5636c17659f2d99583c82ef4ef97f2

I've been looking at the guide here, but can't really get it to work and haven't really used WTF myself so maybe just needs some more thinking.

https://wtforms.readthedocs.io/en/latest/crash_course.html#using-forms


Check this bit's right if the search button takes you nowhere, or back to index.

https://github.com/vulcan25/Code-Bank/blob/master/website2/templates/base_index.html#L53-L57

**Comment by vulcan25** [comment]
sry nuked tht repo, here's the methods:

```
@app.route('/', methods=['GET', 'POST'])
def index():
    search = KinaseForm()

    # if request.method == 'POST':
    #     return search_results(search)

    return render_template('index.html', form=search)

@app.route('/results', methods=['POST'])
def search_results(request):
    if request.method=="POST":
        results = []
        myForm = KinaseForm(request.POST)

        print ('SEARCH: ',myForm.search.data)
        qry = db_session.query(Alt_names).filter(Alt_names.Kinase_name.like('%' + myForm.search.data + '%'))


        results = qry.all()

        if not results:
            abort(404)
        else:
            # display results
            table = Results(results)
            table.border = True
            return render_template('results.html', table=table)
```

**Comment by ishackm** [comment]
Hi, is it working now when u run it ?

**Comment by ishackm** [comment]
Hi, i cant seem to run your repository because i cant find it at all. Can you put it back please?

**Comment by vulcan25** [comment]
Stripped down: https://github.com/vulcan25/code-bank you might want to clone that and copy the static files back in then just ice the `.git` folder and init a new repo for yourself.

**Comment by ishackm** [comment]
Oh ok are the bugs fixed?

**Comment by ishackm** [comment]
Can you search the Kinase?

**Comment by ishackm** [comment]
Hi, when I search for a kinase,

I get the following error:

`TypeError: search_results() missing 1 required positional argument: 'request'`

**Comment by ishackm** [comment]
Hi , to make life easier, i have uploaded the folder "Queries" which is python code. I just need help linking it to flask please?

**Comment by ishackm** [comment]
The code works on the Python terminal but how do I link to the web page ? I would really appreciate the help please because the deadline is due on Monday

**Comment by vulcan25** [comment]
I can't complete your project for you man, I only wanted to help you out to learn flask.  You seem to be doing alright.

I'll give you a pointer.  Ditch POSTing the form, use the get method instead, so in `base_index.html` where the form is defined, the submission `method` should be set to get.

    <form method="get" action="{{ url_for('main_bp.search_results')}}">

Then change the route slightly to handle this:

```
@app.route('/results', methods=['GET'])
def search_results():

    if not request.args.get('search'):
        abort(404) # Maybe make this render a template that tells them the search term was empty
    else:
        print ('Query was:', query)
        # display results
        # table = Results(results)
        # table.border = True
        return render_template('results.html')#table=table)
```

With any luck, when you type something in the search box, it'll display that term in the url like `/results?search=my+sample+search` and you should see the print line print your search terms to the console.

You'd also need to edit the `results.html` template to include the base templates, as that will render blank in the current code.

Then from your queries folder, take the `[in] 7` section from https://github.com/ishackm/Code-Bank/blob/master/Queries/Kinase%20search%20code%201.1.ipynb (for example)

the logic's all there!

You've already got a `db_session` so you might want to change that import line to import it as `session` instead of `db_session` so it works with that code.

Then it's just a case of what you did earlier with the upload code.

Take that `input()` statement out, stick it all in a separate function, and pass `search` in (which is the `query` variable in the route code above.

Then at the end, instead of doing:

```
for x in searchresult:
    print(x)
```

just `return searchresult` meaning from the route you could call it like:

    result = new_function(query)

then hand result to the table rendering code.

Most of this is similar to what I've already been over, so if you were following you should really be able to put it together.

**Comment by vulcan25** [comment]
OF course, in `config.py`:

`ENGINE_STRING = 'sqlite:///test.db'`

must actually point to the correct database.

**Comment by ishackm** [comment]
Apologies for the previous post. I will try my best to make it work. I will commit the changes u recommend. Do you mind looking at them, please? Dw, you wont have to do the whole project, just help me along the way. I will greatly appreciate it😊

**Comment by ishackm** [comment]
Hi, do you mind checking my work please?

**Comment by vulcan25** [comment]
I'm not seeing any updates to the stuff in the `website2` folder.  It's all + 17 hours ago?

**Comment by ishackm** [comment]
Hi i have uploaded website3. Unfortunately i had to remove some files as they were bugging. So for now please run app_terminal_v2.py. i will include the other files later.

**Comment by ishackm** [comment]
index.html
![index_html](https://user-images.githubusercontent.com/45857636/52434127-8f91b580-2b06-11e9-81cf-6d5117b9d437.png)

This is the HTML form, the name of the form is: `kinase_search`

![app_2 py](https://user-images.githubusercontent.com/45857636/52434315-f8792d80-2b06-11e9-9729-4c837de9ea2a.png)

![app py](https://user-images.githubusercontent.com/45857636/52434189-b4862880-2b06-11e9-8160-7cedd890b272.png)

This is the `app_terminal_v2.py`, As you can see I have copied the one for the upload section 

However, when I search something, i get the following error:

![error](https://user-images.githubusercontent.com/45857636/52434444-3aa26f00-2b07-11e9-8ffd-d5e9dcc6b647.png)

Can you help fix me this errror please?
and how can I write in the results.html page so that It renders the list as a table?

Many Thanks,
Ishack

**Comment by vulcan25** [comment]
From that error you should check where the`/kinase_results` route points to which is the function `kinase_results()`.  look towards the end then figure out why the `return would be none` or `missing a return statement`

```
    if likequery!=[]:
        k_alts=[]
        for x in likequery:
            if x.Kinase_name not in k_alts:
                k_alts.append(x.Kinase_name)
        for x in k_alts:
            likequery=session.query(kinase).filter(kinase.c.Kinase_name.like(x)).all()
            if likequery!=[]: #if search found kinases, aka if the search entry was a kinase name
                for x in likequery:
                    if x not in searchresult:
                        searchresult.append(x[:3])
            result= new_function(query)
    elif not request.args.get('search'):
        return "Invalid Kinase Name, Please check again"
    else:
        print ('Query was:',query)
return render_template('kinase_result.html')
```

If the first `if` clause is correct, nothing is returned.   So immediately above the elif line you need to make that return statement.

This doesn't look right though:
`result= new_function(query)`

Looks more like you want to return the `searchresults`.  Be careful of the for loop that you might make that fall into.  The following statement should happen when the above for loop exits, so something like:

```
# [snip]
        for x in k_alts:
            likequery=session.query(kinase).filter(kinase.c.Kinase_name.like(x)).all()
            if likequery!=[]: #if search found kinases, aka if the search entry was a kinase name
                for x in likequery:
                    if x not in searchresult:
                        searchresult.append(x[:3])
        return render_template('kinase_result.html', searchresult = searchresult)
    elif not request.args.get('search'):
# [snip]
```

The syntaxt of that render template line also passes `searchresult` to the template which becomes available as: {{ searchresult }} in the template. 

Then in `kinase_result.html` you should be inheriting from a base template for design purposes, but also do something like this.

```
{% if searchresult %}
<table>
	<tr>
		<td> Kinase name </td>
		<td> Protein name </td>
	</tr>

	{% for result in searchresult %}
	<tr>
		<td> {{ result.Kinase_name }}
		<td> {{ result.Protein_name  }}
	</tr>
	{% endfor %}
</table>
{% else %}
<p> No results found</p>
{% endif %}
```

That way the template logic handles what to display on the frontend, depending on whether anything came back in the search or not.

**Comment by ishackm** [comment]
Thanks, that error is gone but I have a new one:

`

`TypeError: kinase_results() missing 1 required positional argument: 'search'`

app_terminal_v2.py

```
`@app.route('/kinase_results', methods=['GET','POST'])
def kinase_results(search):
    search1="%"+search+"%"
    searchresult=[]
    likequery=session.query(kinase).filter(kinase.c.Kinase_name.like(search1)).all()
    if likequery!=[]: #if search found kinases, aka if the search entry was a kinase name
        for x in likequery:
            if x not in searchresult:
                searchresult.append(x[:3])
    likequery=session.query(kinase).filter(kinase.c.Accession_number.like(search1)).all()
    if likequery!=[]: #if search was an accession number or like an accession number
        for x in likequery:
            if x not in searchresult:
                searchresult.append(x[:3])
    likequery=session.query(kinase).filter(kinase.c.Gene_symbol.like(search1)).all()
    if likequery!=[]: #if search was a gene symbol or like a gene symbol
        for x in likequery:
            if x not in searchresult:
                searchresult.append(x[:3])

    likequery=session.query(alt_names).filter(alt_names.c.Protein_name.like(search1)).all()
    if likequery!=[]:
        k_alts=[]
        for x in likequery:
            if x.Kinase_name not in k_alts:
                k_alts.append(x.Kinase_name)
        for x in k_alts:
            likequery=session.query(kinase).filter(kinase.c.Kinase_name.like(x)).all()
            if likequery!=[]: #if search found kinases, aka if the search entry was a kinase name
                for x in likequery:
                    if x not in searchresult:
                        searchresult.append(x[:3])
        return render_template('kinase_result.html', searchresult = searchresult)
    elif not request.args.get('search'):
        return "Kinase not found, please check again "
`
```

What am I doing wrong  here, please?

**Comment by vulcan25** [comment]
Looks like the way you're calling the function is wrong:

```
@app.route('/kinase_results', methods=['GET','POST']) ``
    def kinase_results(search):
```

You're passing the `search` variable in with a `GET` method, so you need to obtain it like this.

```
    @app.route('/kinase_results', methods=['GET','POST']) ``
    def kinase_results():
        search = request.args.get('search'):
        # Rest of code
        search1="%"+search+"%"
        searchresult=[]
        # etc, etc
```


Interestingly, for future reference, another way to get a value in to there is as a URL path, so you can do:

```
    @app.route('/wherever/<string:item>/<int:count>', methods=['GET','POST']) ``
    def kinase_results(item, count):
        print ('The item is: ', item)
        print ('The count is: ', count)
```
`/whatever/laptop/78` is then a valid URL and you can process each variable accordingly.

As you can see in the [docs](http://flask.pocoo.org/docs/1.0/quickstart/#variable-rules) there are a number of different types supported.

**Comment by ishackm** [comment]
Hi, thanks but sorry it shows the following error:

`    search = request.args.get('search'):`

`SyntaxError: invalid syntax`

**Comment by vulcan25** [comment]
You sure it relates to that line, can you post the full traceback.

**Comment by vulcan25** [comment]
sorry the colon lol, get rid of that at the end.

**Comment by ishackm** [comment]
Thanks but sorry I got a new error now lol, I am sorry for getting these  annoying bugs


`

`TypeError: can only concatenate str (not "NoneType") to str`
`

app_terminal_v2.py

```

@app.route('/kinase_results', methods=['GET','POST']) 
def kinase_results():
    search = request.args.get('search')
    search1="%"+search+"%"
    searchresult=[]
    likequery=session.query(kinase).filter(kinase.c.Kinase_name.like(search1)).all()
    if likequery!=[]: #if search found kinases, aka if the search entry was a kinase name
        for x in likequery:
            if x not in searchresult:
                searchresult.append(x[:3])
    likequery=session.query(kinase).filter(kinase.c.Accession_number.like(search1)).all()
    if likequery!=[]: #if search was an accession number or like an accession number
        for x in likequery:
            if x not in searchresult:
                searchresult.append(x[:3])
    likequery=session.query(kinase).filter(kinase.c.Gene_symbol.like(search1)).all()
    if likequery!=[]: #if search was a gene symbol or like a gene symbol
        for x in likequery:
            if x not in searchresult:
                searchresult.append(x[:3])

    likequery=session.query(alt_names).filter(alt_names.c.Protein_name.like(search1)).all()
    if likequery!=[]:
        k_alts=[]
        for x in likequery:
            if x.Kinase_name not in k_alts:
                k_alts.append(x.Kinase_name)
        for x in k_alts:
            likequery=session.query(kinase).filter(kinase.c.Kinase_name.like(x)).all()
            if likequery!=[]: #if search found kinases, aka if the search entry was a kinase name
                for x in likequery:
                    if x not in searchresult:
                        searchresult.append(x[:3])
        return render_template('kinase_result.html', searchresult = searchresult)
    elif not request.args.get('search'):
        return "Kinase not found, please check again "
```

**Comment by vulcan25** [comment]
Does it work if you specify the argument in the URL?  `/kinase_results?search=hello` 

If it does, I'm guessing this bug is due to:

    search1="%"+search+"%"

Which is `none` if no search term is entered.

The solution would be to cut that `elif` line from the end, and stick it in at the top:
```
def kinase_results():
    if not request.args.get('search'):
        return render_template('kinase_result.html')
    search = request.args.get('search')
    search1="%"+search+"%"
    searchresult=[]
```

That way if `search` is none, it will render the template immedaitely (In Python, a `return` statement inside a function being triggered stops the rest of the code from executing).

Otherwise it will continue to  `search = request.args.get('search')` which will definately have a value.

Validation should probably be done somewhere hereish.

**Comment by ishackm** [comment]
Hi, sorry I dont seem to understand , can you post the full code please. Also the URL method does not work
